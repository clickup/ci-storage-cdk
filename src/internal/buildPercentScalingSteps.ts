type PercentScalingStep = {
  pctTarget: number;
  upper?: number;
  lower?: number;
  change: number;
  pctAfter: number;
};

/**
 * Builds scaling steps, so that the further the actual percentage metric is
 * from pctTarget, the more instances will be added or removed (so the resulting
 * instances busy percentage will remain around pctTarget if nothing changes).
 */
export function buildPercentScalingSteps(
  pctTarget: number,
  steps: number,
): PercentScalingStep[] {
  const result: PercentScalingStep[] = [];

  // An example number of instances to illustrate the "after" situation.
  const N = 100;

  {
    const stepSize = pctTarget / steps;
    for (let i = steps; i >= 1; i--) {
      const lower = Math.round(pctTarget - i * stepSize);
      const upper = Math.round(pctTarget - (i - 1) * stepSize);
      const change = Math.round((upper / pctTarget - 1) * 100);
      const pctAfter = Math.round(
        ((N * upper * 0.01) / (N + N * change * 0.01)) * 100,
      );
      result.push({
        pctTarget,
        lower: lower <= 0 ? undefined : lower,
        upper,
        change,
        pctAfter,
      });
    }
  }

  {
    const stepSize = (100 - pctTarget) / steps;
    for (let i = 1; i <= steps; i++) {
      const lower = Math.round(pctTarget + (i - 1) * stepSize);
      const upper = Math.round(pctTarget + i * stepSize);
      const change = Math.round((upper / pctTarget - 1) * 100);
      const pctAfter = Math.round(
        ((N * upper * 0.01) / (N + N * change * 0.01)) * 100,
      );
      result.push({
        pctTarget,
        lower,
        upper: upper >= 100 ? undefined : upper,
        change,
        pctAfter,
      });
    }
  }

  return result;
}
