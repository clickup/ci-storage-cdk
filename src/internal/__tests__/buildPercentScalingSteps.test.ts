import { inspect } from "util";
import { buildPercentScalingSteps } from "../buildPercentScalingSteps";

test("buildPercentScalingSteps", () => {
  expect(
    inspect(buildPercentScalingSteps(70, 4), { compact: true }),
  ).toMatchSnapshot();
  expect(
    inspect(buildPercentScalingSteps(50, 4), { compact: true }),
  ).toMatchSnapshot();
});
