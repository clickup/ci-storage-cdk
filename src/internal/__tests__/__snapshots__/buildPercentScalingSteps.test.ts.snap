// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildPercentScalingSteps 1`] = `
"[ { pctTarget: 70, lower: undefined, upper: 18, change: -74, pctAfter: 69 },
  { pctTarget: 70, lower: 18, upper: 35, change: -50, pctAfter: 70 },
  { pctTarget: 70, lower: 35, upper: 53, change: -24, pctAfter: 70 },
  { pctTarget: 70, lower: 53, upper: 70, change: 0, pctAfter: 70 },
  { pctTarget: 70, lower: 70, upper: 78, change: 11, pctAfter: 70 },
  { pctTarget: 70, lower: 78, upper: 85, change: 21, pctAfter: 70 },
  { pctTarget: 70, lower: 85, upper: 93, change: 33, pctAfter: 70 },
  { pctTarget: 70, lower: 93, upper: undefined, change: 43, pctAfter: 70 } ]"
`;

exports[`buildPercentScalingSteps 2`] = `
"[ { pctTarget: 50, lower: undefined, upper: 13, change: -74, pctAfter: 50 },
  { pctTarget: 50, lower: 13, upper: 25, change: -50, pctAfter: 50 },
  { pctTarget: 50, lower: 25, upper: 38, change: -24, pctAfter: 50 },
  { pctTarget: 50, lower: 38, upper: 50, change: 0, pctAfter: 50 },
  { pctTarget: 50, lower: 50, upper: 63, change: 26, pctAfter: 50 },
  { pctTarget: 50, lower: 63, upper: 75, change: 50, pctAfter: 50 },
  { pctTarget: 50, lower: 75, upper: 88, change: 76, pctAfter: 50 },
  { pctTarget: 50, lower: 88, upper: undefined, change: 100, pctAfter: 50 } ]"
`;
